sirepo/auth/ldap.py

```diff
@@ -0,0 +1,87 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2018-2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+# TODO(BG): Write test
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+
+        # validates username/password, escapes special chars for ldap dn
+        def _validate_escape_credentials(req):
+            email = req.req_data.email.lower()
+            if len(email) > 256 or len(req.req_data.password) > 256:
+                raise sirepo.util.UserAlert(
+                    "invalid user and/or password",
+                    f"email={email} or password greater than 256 chars",
+                )
+            if not email or not req.req_data.password:
+                raise sirepo.util.UserAlert(
+                    "invalid user and/or password",
+                    f"email={email} or password is none/zero length",
+                )
+            return (
+                "mail=" + re.sub(_ESCAPE_DN_MAIL, r"\\\1", email) + _cfg.base_dn,
+                email,
+                req.req_data.password,
+            )
+
+        def _authorize_ldap(dn, email, p):
```

Comment:
`_authorize_ldap` is redundant. Call this `_bind`, since it is an internal function that is describing what it is doing.

The parameter `p` is too short. Call it password. However, you will instead call a single argument here `creds` which is returned from `_validate_and_escape_creds`.